interface/grpc/core/list_services.go

```diff
@@ -3,14 +3,53 @@ package core
 import (
 	"context"
 
+	"github.com/mesg-foundation/core/api"
 	"github.com/mesg-foundation/core/protobuf/coreapi"
+	service "github.com/mesg-foundation/core/service"
+	"github.com/mesg-foundation/core/x/xerrors"
 )
 
 // ListServices lists services.
 func (s *Server) ListServices(ctx context.Context, request *coreapi.ListServicesRequest) (*coreapi.ListServicesReply, error) {
-	services, err := s.api.ListServices()
+	var filters []api.ListServicesFilter
+	if request.FilterActive {
+		filters = append(filters, api.ListRunningServicesFilter())
+	}
+	services, err := s.api.ListServices(filters...)
 	if err != nil {
 		return nil, err
 	}
-	return &coreapi.ListServicesReply{Services: toProtoServices(services)}, nil
+
+	protoServices := toProtoServices(services)
+
+	var (
+		servicesLen = len(services)
+		errC        = make(chan error, servicesLen)
+	)
+
+	// fill services status info.
+	for _, s := range services {
+		go func(s *service.Service) {
+			status, err := s.Status()
+			if err == nil {
+				for _, ss := range protoServices {
+					if ss.ID == s.ID {
+						ss.Status = toProtoServiceStatusType(status)
+					}
+				}
+			}
+			errC <- err
+		}(s)
+
+	}
+
+	var errs xerrors.Errors
+
+	for i := 0; i < servicesLen; i++ {
```

Comment:
+1 for the wait group, I think it makes the code more readable than a for that is waiting for some events in the chan.

@krhubert I don't get the difference between the buffered and unbuffered version. I would go myself on the first one, what will be the advantage of using the second one ?

Reply:
thanks, i cannot see an advantage of using _WaitGroup_ here except using _range_ syntax while receiving from _errC_ which makes readability slightly better. also check std package [example](https://github.com/golang/go/blob/master/src/net/http/transport.go#L1175-L1177).

lets give a decision about which way to go. i'm neutral about this, i see that using _range_ has an upside for readability but it also introduces more code.

Reply:
The example you gave it's slightly different. Here they don't wait for goroutines to finish but provide [timeout](https://github.com/golang/go/blob/master/src/net/http/transport.go#L1164) for TLSHandshake execution. WIthout timeout, there will be no point to put TLSHandshake inside goroutine.

Reply:
I also suggest to use a WaitGroup to close the chan and use a simple `range` to iterate over the errors.
I suggest to avoid the classic `for i := 0; i < len(arr); i++ {` and anything that directly deal with index on array. It's a source of error.

Reply:
do three of you have an agreement on this? please like this comment, i'll switch to _WaitGroup_ if i see likes >= 2. :)

Reply:
About the second example, the code is inside the test package and it's quite old. 

```
8e69d43b32b src/net/net_test.go     (Brad Fitzpatrick 2016-09-27 20:50:57 +0000 451)    for i := 0; i < 2; i++ {
8e69d43b32b src/net/net_test.go     (Brad Fitzpatrick 2016-09-27 20:50:57 +0000 452)            if err := <-errc; err != nil {
8e69d43b32b src/net/net_test.go     (Brad Fitzpatrick 2016-09-27 20:50:57 +0000 453)                    t.Fatal(err)
8e69d43b32b src/net/net_test.go     (Brad Fitzpatrick 2016-09-27 20:50:57 +0000 454)            }
8e69d43b32b src/net/net_test.go     (Brad Fitzpatrick 2016-09-27 20:50:57 +0000 455)    }
```

Maybe there are no strict reviews for tests or maybe you can allow a little bit not standard code in tests. I don't know, but ...

Such cases of errc usage are only 20 in the whole codebase (and 14 of them are for tests).

```
➜  net git:(master) gogrep 'err := <-err' | wc -l      
20
➜  net git:(master) gogrep 'err := <-err' | grep _test.go | wc -l
14
```

It may happen for everyone to produce not "optimal" code. And about your code - the solution is perfectly fine and it works. But it's not the go way - what I mean by that - over the time go developers used to some patterns. WaitGroup is one of it. It just feels more natural for others if they see the pattern like this because everyone used to it and recognize it.

The other example is much simpler from today:

```go
const (
  A int = 1
  B int = 2
)
```

The code works and it's nothing wrong with it, but it looks wired because everyone used to

```go
const (
  A int = 1
  B       = 2
)
````

There are some patterns that everyone understands and it's preferable to use them (for example `Option` for optional arguments :))

Reply:
thank you for the research! i still think that using err chans to wait end of the goroutines is an anti-pattern but i'm ok with the two ways.